### PR TITLE
add restart modules

### PR DESCRIPTION
**Pull request by juliangruber**
Alternative implementation to https://github.com/filecoin-station/core/pull/169.

Notable differences:

Instead of the zinnia lib restarting itself internally, we make `zinnia.start()` throw an error when it exists. Then, we use `p-retry()` to take care of restarting it

### CODE REVIEW

File: commands/station.js [hunk 1]
```diff
@@ -64,7 +65,7 @@ export const station = async ({ json, experimental }) => {
         })
       },
       onMetrics: m => metrics.submit('zinnia', m)
-    })
+    }), { retries: 1000 })
```
Comment (juliangruber) [inline commands/station.js:68]: I tried `Infinity`, unfortunately that doesn't work. See https://github.com/sindresorhus/p-retry/issues/58

Comment (bajtos) [reply commands/station.js:68]: Can we use something like a max-safe integer?

File: commands/station.js [hunk 2]
```diff
@@ -50,7 +51,7 @@ export const station = async ({ json, experimental }) => {
   })
 
   const modules = [
-    zinniaRuntime.start({
+    pRetry(() => zinniaRuntime.start({
```
Comment (bajtos) [inline commands/station.js:54]: With this new implementation, I think the name `start` is no longer accurate. What do you think about renaming `start` to `run`?

```suggestion
    pRetry(() => zinniaRuntime.run({
```

Comment (juliangruber) [reply commands/station.js:54]: Good call, yes!

File: lib/zinnia.js [hunk 3]
```diff
@@ -72,29 +79,30 @@ export async function start ({
...
-  } catch (err) {
-    const errorMsg = err instanceof Error ? err.message : '' + err
-    const message = `Cannot start Zinnia: ${errorMsg}`
-    onActivity({ type: 'error', message })
-  }
+  await Promise.all([
+    (async () => {
+      try {
+        await Promise.race([
+          readyPromise,
```
Comment (juliangruber) [inline lib/zinnia.js:86]: remove `readyPromise`

File: commands/station.js
Comment (juliangruber) [inline commands/station.js:1]: do the same thing for `bacalhau`

File: lib/zinnia.js [hunk 4]
```diff
@@ -57,6 +58,12 @@ export async function start ({
     }
   })
 
+  childProcess.on('exit', (code, signal) => {
```
Comment (juliangruber) [inline lib/zinnia.js:61]: move this back below

File: lib/zinnia.js (hunk 3 above)
Comment (juliangruber) [reply lib/zinnia.js:86]:
```ja
await childProcess
throw new Error()
```

File: commands/station.js (hunk 1 above)
Comment (juliangruber) [reply commands/station.js:68]: it also fails with that unfortunately

**Review by bajtos — APPROVED** [review]

File: lib/zinnia.js [hunk 5]
```diff
@@ -57,44 +57,38 @@ export async function start ({
...
-    childProcess.stdout.on('data', data => {
-      handleEvents({ onActivity, onMetrics, text: data })
-    })
-    childProcess.stderr.pipe(process.stderr, { end: false })
-
-    childProcess.stdout.once('data', _data => {
-      // This is based on an implicit assumption that zinniad reports an info activity
-      // after it starts
-      resolve()
-    })
```
[the remark this answers is not in the document]
Comment (juliangruber) [reply lib/zinnia.js:71]: That is correct. It was the outcome of a meeting by the two of us that this logic can be dropped. I agree, let's see how this goes

**Pull request merged**